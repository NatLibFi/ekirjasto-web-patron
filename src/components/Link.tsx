/** @jsx jsx */
import { jsx } from "theme-ui";
import * as React from "react";
import BaseLink from "next/link";
import useLinkUtils, { LinkUtils } from "./context/LinkUtilsContext";

type CollectionLinkProps = {
  collectionUrl: string;
};
type BookLinkProps = {
  bookUrl: string;
};
type OtherLinkProps = {
  href: string;
};

type BaseLinkProps = Omit<
  React.ComponentProps<typeof BaseLink>,
  "href" | "as"
> &
  Omit<React.ComponentPropsWithoutRef<"a">, "href"> & {
    className?: string;
  };

export type LinkProps = BaseLinkProps &
  (CollectionLinkProps | BookLinkProps | OtherLinkProps);

/**
 * converts bookUrl and collectionUrl to as/href props
 * prepends with multi library path if needed
 * removes consumed props and returns normalized props
 */
const buildLinkFromProps = (props: LinkProps, linkUtils: LinkUtils) => {
  if ("bookUrl" in props) {
    const { bookUrl, ...rest } = props;
    return { href: linkUtils.buildBookLink(bookUrl), ...rest };
  }
  if ("collectionUrl" in props) {
    const { collectionUrl, ...rest } = props;
    return { href: linkUtils.buildCollectionLink(collectionUrl), ...rest };
  }
  const { href, ...rest } = props;
  return {
    href: linkUtils.buildMultiLibraryLink(props.href),
    ...rest
  };
};
/**
 * Extends next/Link to:
 *  - add styles
 *  - automatically prepend the library ID if using multiple libraries.
 *  - allows user to pass in ONE OF:
 *    - "href" and "as", a normal next/Link
 *    - "bookUrl"
 *    - "collectionUrl"
 */
const Link = React.forwardRef<HTMLAnchorElement, LinkProps>(
  ({ children, className, ...props }, ref) => {
    const linkUtils = useLinkUtils();
    const {
      href,
      prefetch,
      replace,
      scroll,
      shallow,
      ...rest
    } = buildLinkFromProps(props, linkUtils);
    return (
      <BaseLink
        href={href}
        prefetch={prefetch}
        replace={replace}
        scroll={scroll}
        shallow={shallow}
        passHref
      >
        <a
          ref={ref}
          sx={{
            textDecoration: "none",
            color: "inherit",
            "&:hover": { color: "inherit", textDecoration: "underline" }
          }}
          className={className}
          {...rest}
        >
          {children}
        </a>
      </BaseLink>
    );
  }
);

export default Link;
