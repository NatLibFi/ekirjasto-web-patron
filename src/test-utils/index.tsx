import * as React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom";
import ContextProvider from "../components/context/ContextProvider";
import Adapter from "enzyme-adapter-react-16";
import { configure } from "enzyme";
import library from "./fixtures/library";
import BasicAuthPlugin from "../auth/basicAuthPlugin";
import buildStore from "owc/store";
import { State } from "owc/state";
import { LibraryData } from "../interfaces";
import "./mockScrollTo";
import * as fixtures from "./fixtures";
import ActionsCreator from "owc/actions";
import DataFetcher from "owc/DataFetcher";
import { adapter } from "owc/OPDSDataAdapter";
import serializer from "jest-emotion";
import { MockNextRouterContextProvider } from "./mockNextRouter";
import { NextRouter } from "next/router";
import { enableFetchMocks } from "jest-fetch-mock";
import setEnv from "./setEnv";

enableFetchMocks();
expect.addSnapshotSerializer(serializer);

export { fixtures, setEnv };

// configure the enzyme adapter
configure({ adapter: new Adapter() });

/**
 * mock out the window.URL.createObjectURL since it isn't
 * available on jsdom
 */
const mockCreateObjectURL = jest.fn();
(global as any)["URL"].createObjectURL = mockCreateObjectURL;

/**
 * We create the actions and fetcher here so that they can
 * be imported by our test file and spied on using:
 * jest.spyOn(actions, "fetchSearchDescription")
 * and then asserted against
 */
export const fetcher = new DataFetcher({ adapter });
export const actions = new ActionsCreator(fetcher);

type CustomRenderOptions = Parameters<typeof render>[1] & {
  router?: Partial<NextRouter>;
  initialState?: State;
  library?: LibraryData;
  colors?: { primary: string; secondary: string };
};
const customRender = (ui: any, options?: CustomRenderOptions) => {
  const pathFor = jest
    .fn()
    .mockImplementation((collectionUrl?: string, bookUrl?: string) =>
      `/${collectionUrl}` + bookUrl ? `/${bookUrl}` : ""
    );

  const store = buildStore(options?.initialState, [BasicAuthPlugin], pathFor);

  // spy on dispatch by default
  const origDispatch = store.dispatch;
  const mockDispatch: jest.Mock<typeof origDispatch> = jest
    .fn()
    .mockImplementation(origDispatch);
  store.dispatch = mockDispatch as typeof origDispatch;

  const AllTheProviders: React.FC = ({ children }) => {
    return (
      <MockNextRouterContextProvider router={options?.router}>
        <ContextProvider
          library={options?.library ?? library}
          store={store}
          fetcher={fetcher}
          actions={actions}
        >
          {children}
        </ContextProvider>
      </MockNextRouterContextProvider>
    );
  };

  return {
    ...render(ui, { wrapper: AllTheProviders, ...options }),
    // we pass our mocks along so they can be used in assertions
    store,
    dispatch: mockDispatch
  };
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
