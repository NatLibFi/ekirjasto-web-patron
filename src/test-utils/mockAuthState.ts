import { AuthCredentials, OPDS1 } from "interfaces";
import Cookie from "js-cookie";

const mockCookie = Cookie as any;

export const creds: AuthCredentials = {
  token: "some-token",
  methodType: OPDS1.BasicAuthType
};

const str = JSON.stringify;

export default function mockAuthenticatedOnce(
  credentials: AuthCredentials | null = creds
) {
  mockCookie.get.mockReturnValueOnce(str(credentials));
}

export function mockAuthenticated(credentials: AuthCredentials | null = creds) {
  mockCookie.get.mockReturnValue(str(credentials));
}
