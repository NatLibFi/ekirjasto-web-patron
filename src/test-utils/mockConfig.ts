import * as configModule from "config";
import { AppConfig } from "interfaces";

export default function mockConfig(custom?: Partial<AppConfig>) {
  (configModule.APP_CONFIG as any) = {
    ...defaultMock,
    ...custom
  };
}

const defaultMock: AppConfig = {
  bugsnagApiKey: null,
  gtmId: null,
  companionApp: "simplye",
  showMedium: true,
  libraries: {
    lib1: "http://lib1.com"
  },
  mediaSupport: {
    "application/epub+zip": "show",
    "application/kepub+zip": "show",
    "application/pdf": "show",
    "application/x-mobipocket-ebook": "show",
    "application/x-mobi8-ebook": "show",
    "application/vnd.overdrive.circulation.api+json;profile=ebook": "show",
    // # External read online type (Like Overdrive)
    'text/html;profile="http://librarysimplified.org/terms/profiles/streaming-media"':
      "show",
    // # AxisNow document (read online in Webpub Viewer)
    "application/vnd.librarysimplified.axisnow+json": "show",
    // # Audiobooks
    "application/audiobook+json": "redirect",
    "application/vnd.overdrive.circulation.api+json;profile=audiobook":
      "redirect",
    // # INDIRECT TYPES
    // # OPDS Entry Indirection type
    "application/atom+xml;type=entry;profile=opds-catalog": {
      'text/html;profile="http://librarysimplified.org/terms/profiles/streaming-media"':
        "show"
    },
    // # Adobe Encryption
    "application/vnd.adobe.adept+xml": {
      "application/epub+zip": "redirect-and-show"
    }
    // # Bearer Token Exchange
    // # application/vnd.librarysimplified.bearer-token+json:
  }
};
