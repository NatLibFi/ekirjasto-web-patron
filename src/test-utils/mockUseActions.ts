import { mockDispatch } from "./index";
import ActionsCreator from "opds-web-client/lib/actions";
import DataFetcher from "opds-web-client/lib/DataFetcher";
import { useActions } from "opds-web-client/lib/components/context/ActionsContext";
/**
 * First we mock the ActionsCreator
 */
jest.mock("opds-web-client/lib/actions");
const MockedActionCreator = ActionsCreator as jest.MockedClass<
  typeof ActionsCreator
>;

/**
 * We should also mock the DataFetcher but this is currently causing
 * an issue because jest reads the DataFetcher mock from opds-web-client
 * which doest work here
 */

export const mockFetcher = new DataFetcher();

// const mockedUseActions = useActions as jest.Mock;
export const mockActionCreator = new MockedActionCreator(new DataFetcher());

const mockUseActions = jest.fn().mockReturnValue({
  actions: mockActionCreator,
  fetcher: mockFetcher,
  dispatch: mockDispatch
});
jest.mock("opds-web-client/lib/components/context/ActionsContext", () => ({
  ...jest.requireActual(
    "opds-web-client/lib/components/context/ActionsContext"
  ),
  useActions: mockUseActions
}));

beforeEach(() => {
  MockedActionCreator.mockClear();
  mockUseActions.mockClear();
  mockDispatch.mockClear();
});
