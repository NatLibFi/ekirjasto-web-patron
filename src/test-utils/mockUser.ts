import * as user from "components/context/UserContext";

const useUserSpy = jest.spyOn(user, "default");

const defaultUser: ReturnType<typeof user.default> = {
  isAuthenticated: true,
  loans: [],
  isLoading: false,
  status: "authenticated",
  refetchLoans: jest.fn(),
  signIn: jest.fn(),
  signOut: jest.fn(),
  setBook: jest.fn(),
  error: undefined,
  token: "user-token"
};

export default function mockUser(data: Partial<typeof defaultUser> = {}) {
  const user = {
    ...defaultUser,
    ...data
  };
  useUserSpy.mockReturnValue(user);
}
