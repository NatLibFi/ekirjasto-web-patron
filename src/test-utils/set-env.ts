import { config } from "test-utils/fixtures/config";

process.env.APP_CONFIG = JSON.stringify(config);
