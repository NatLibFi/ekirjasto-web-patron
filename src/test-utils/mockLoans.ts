import { BookData } from "interfaces";
import * as fetch from "dataflow/opds1/fetch";
import { fixtures } from "test-utils";

(fetch as any).fetchCollection = jest.fn();
export const mockedFetchCollection = fetch.fetchCollection as jest.MockedFunction<
  typeof fetch.fetchCollection
>;

export default function mockLoans(books: BookData[] = fixtures.loans.books) {
  mockedFetchCollection.mockResolvedValue({ ...fixtures.loans, books });
}
