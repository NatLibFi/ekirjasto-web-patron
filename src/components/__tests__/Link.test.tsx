import * as React from "react";
import { render } from "../../test-utils";
import Link from "../Link";

test("Renders expected styles", () => {
  const utils = render(<Link href="/somewhere">click here</Link>);
  expect(utils.getByText("click here")).toMatchSnapshot();
});
test("Renders proper href with standard next.js href prop", () => {
  const utils = render(<Link href="/somewhere-new">click here</Link>);
  expect(utils.getByText("click here")).toHaveAttribute(
    "href",
    "/somewhere-new"
  );
});
test("Renders proper href with standard next.js href+as props", () => {
  const utils = render(
    <Link href="/[library]/home" as="/mylib/home">
      click here
    </Link>
  );
  expect(utils.getByText("click here")).toHaveAttribute("href", "/mylib/home");
});
test("Renders proper href with bookUrl prop", () => {
  const utils = render(<Link bookUrl="http://some.book.com/">click here</Link>);
  expect(
    utils.getByText("click here").closest("a")?.href
  ).toMatchInlineSnapshot(
    `"http://test-domain.com/book/http%3A%2F%2Fsome.book.com%2F"`
  );
});
test("Renders proper href with collectionUrl prop", () => {
  const utils = render(
    <Link collectionUrl="http://some.collection.com/">click here</Link>
  );
  expect(
    utils.getByText("click here").closest("a")?.href
  ).toMatchInlineSnapshot(
    `"http://test-domain.com/collection/http%3A%2F%2Fsome.collection.com"`
  );
});
test("When collectionUrl is your base url, links to home", () => {
  const utils = render(
    <Link collectionUrl="http://test-cm.com/catalogUrl">click here</Link>
  );
  expect(utils.getByText("click here").closest("a")?.href).toEqual(
    "http://test-domain.com/"
  );
});
