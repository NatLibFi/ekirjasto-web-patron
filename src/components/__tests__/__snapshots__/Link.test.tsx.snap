// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders expected styles 1`] = `
.emotion-0 {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
}

.emotion-0:hover {
  color: inherit;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<a
  class="emotion-0"
  href="/somewhere"
>
  click here
</a>
`;
